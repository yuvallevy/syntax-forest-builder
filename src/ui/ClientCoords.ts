type ClientX = number;
type ClientY = number;

type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientRect = {
  topLeft: ClientCoords,
  bottomRight: ClientCoords,
};

export default ClientCoords;
