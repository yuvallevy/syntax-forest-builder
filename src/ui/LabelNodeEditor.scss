.LabelNodeEditorInput {
  position: absolute;
  width: 2em;
  border: 1px solid #ccc;
  border-radius: 3px;
  font-family: 'Times New Roman', Times, serif;
  font-size: 16px;
  text-align: center;
  transform: translate(-50%, -100%);
}
