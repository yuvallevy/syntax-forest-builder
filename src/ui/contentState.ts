import {
  Id, IdMap, StringSlice, Sentence, UnpositionedPlot, UnpositionedTree, TreeAndNodeId, PlotCoordsOffset
} from '../core/types';
import { deleteNodesInTree, InsertedNode, insertNodeIntoTree, transformNodeInTree } from '../mantle/manipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../mantle/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { omitKey } from '../core/objTransforms';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'insertNode', plotId: Id, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotId: Id, nodes: TreeAndNodeId[] }
  | { type: 'setNodeLabel', plotId: Id, node: TreeAndNodeId, newLabel: string }
  | { type: 'setSentence', plotId: Id, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotId: Id, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotId: Id, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setTree', plotId: Id, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotId: Id, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotId: Id, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'insertNode':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotId].trees[action.treeId]),
      };
    case 'deleteNodes':
      const treeId = action.nodes[0].treeId;  // TODO: Use all trees
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: treeId,
        old: state.plots[action.plotId].trees[treeId],
        new: deleteNodesInTree(action.nodes.map(({ nodeId }) => nodeId))(state.plots[action.plotId].trees[treeId]),
      };
    case 'setNodeLabel':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.node.treeId,
        old: state.plots[action.plotId].trees[action.node.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.node.nodeId)(
          state.plots[action.plotId].trees[action.node.treeId]),
      };
    case 'setSentence':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotId].trees[action.treeId]),
      };
    case 'addTree':
      return {
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    case 'removeTree':
      return {
        ...action,
        removedTree: state.plots[action.plotId].trees[action.treeId],
      };
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.treeId]: action.new,
            },
          },
        },
      };
    case 'addTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.newTreeId]: action.newTree,
            },
          },
        },
      };
    case 'removeTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: omitKey(state.plots[action.plotId].trees, action.treeId)
          },
        },
      };
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setTree':
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    case 'addTree':
      return {
        ...action,
        type: 'removeTree',
        plotId: action.plotId,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    case 'removeTree':
      return {
        ...action,
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    default:
      return action;
  }
};

export type ContentState = {
  plots: IdMap<UnpositionedPlot>;
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: {
    'plot': {
      trees: {},
    },
  },
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(reverseUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
