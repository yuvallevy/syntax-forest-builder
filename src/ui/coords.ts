import { PlotCoords, PositionedNode, PositionedTree } from '../content/positioned/types';

type ClientX = number;
type ClientY = number;

export type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientRect = {
  topLeft: ClientCoords;
  bottomRight: ClientCoords;
};

export type PlotRect = {
  topLeft: PlotCoords;
  bottomRight: PlotCoords;
};

// Calculate plot coords from client coords & vice versa, when they're different (when zoom and/or pan are implemented)

export const plotCoordsToClientCoords = (plotCoords: PlotCoords): ClientCoords => ({
  clientX: plotCoords.plotX,
  clientY: plotCoords.plotY,
});

export const clientCoordsToPlotCoords = (clientCoords: ClientCoords): PlotCoords => ({
  plotX: clientCoords.clientX,
  plotY: clientCoords.clientY,
});

export const clientRectToPlotRect = (clientRect: ClientRect): PlotRect => ({
  topLeft: clientCoordsToPlotCoords(clientRect.topLeft),
  bottomRight: clientCoordsToPlotCoords(clientRect.bottomRight),
});

/**
 * Returns the center coordinate of the given node in the given tree with respect to its containing plot.
 */
export const calculateNodeCenterOnPlot = (tree: PositionedTree) => (node: PositionedNode): PlotCoords => ({
  plotX: tree.position.plotX + node.position.treeX,
  plotY: tree.position.plotY + node.position.treeY - 9,
});
