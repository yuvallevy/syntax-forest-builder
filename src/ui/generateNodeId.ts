const ID_TIME_LENGTH = 6;
const ID_RANDOM_LENGTH = 4;
const ID_BASE = 36;

const maxTimestamp = parseInt('1' + '0'.repeat(ID_TIME_LENGTH), ID_BASE);
const maxRandom = parseInt('1' + '0'.repeat(ID_RANDOM_LENGTH), ID_BASE);

const padLeft = (str: string, char: string, minLength: number) =>
  str.length >= minLength ? str
    : (char.repeat(minLength - str.length) + str);

const generateNodeId = () =>
  padLeft((new Date().valueOf() % maxTimestamp).toString(ID_BASE), '0', ID_TIME_LENGTH) +
  padLeft(Math.floor(Math.random() * maxRandom).toString(ID_BASE), '0', ID_RANDOM_LENGTH);

export default generateNodeId;
