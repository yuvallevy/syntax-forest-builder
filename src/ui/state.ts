import { Id, IdMap, NodeSlice, Sentence, UnpositionedPlot, UnpositionedTree } from '../core/types';
import { deleteNodesInTree, InsertedNode, insertNodeIntoTree, transformNodeInTree } from '../mantle/manipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../mantle/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';

export type TreeAndNodeId = {
  treeId: Id;
  nodeId: Id;
};

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'insertNode', plotId: Id, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotId: Id, nodes: TreeAndNodeId[] }
  | { type: 'setNodeLabel', plotId: Id, node: TreeAndNodeId, newLabel: string }
  | { type: 'setSentence', plotId: Id, treeId: Id, newSentence: Sentence, oldSelection: NodeSlice }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | {
    type: 'setTree',
    plotId: Id,
    treeId: Id,
    old: UnpositionedTree,
    new: UnpositionedTree,
  }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'insertNode':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotId].trees[action.treeId]),
      };
    case 'deleteNodes':
      const treeId = action.nodes[0].treeId;  // TODO: Use all trees
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: treeId,
        old: state.plots[action.plotId].trees[treeId],
        new: deleteNodesInTree(action.nodes.map(({ nodeId }) => nodeId))(state.plots[action.plotId].trees[treeId]),
      };
    case 'setNodeLabel':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.node.treeId,
        old: state.plots[action.plotId].trees[action.node.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.node.nodeId)(
          state.plots[action.plotId].trees[action.node.treeId]),
      };
    case 'setSentence':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelection)(
          state.plots[action.plotId].trees[action.treeId]),
      }
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.treeId]: action.new,
            },
          },
        },
      };
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setTree':
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    default:
      return action;
  }
};

export type ContentState = {
  plots: IdMap<UnpositionedPlot>;
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: {
    'plot': {
      trees: {
        'aa': {
          sentence: 'Noun verbs very adverbly.',
          nodes: {
            // 'a': {
            //   label: 'S', offset: { dTreeX: 0, dTreeY: 0 }, children: {
                'b': {
                  label: 'NP', offset: { dTreeX: 0, dTreeY: -10 }, children: {
                    'c': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
                  }
                },
                'd': {
                  label: 'VP', offset: { dTreeX: 0, dTreeY: 0 }, children: {
                    'e': { label: 'V', offset: { dTreeX: 0, dTreeY: 0 }, slice: [5, 10], triangle: false },
                    'f': { label: 'AdvP', offset: { dTreeX: 0, dTreeY: 0 }, slice: [11, 24], triangle: true },
                  }
                },
            //   }
            // },
          },
          offset: { dPlotX: 200, dPlotY: 250 },
        },
      },
    },
  },
};

export const undoableInitialState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const undoableReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(reverseUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
