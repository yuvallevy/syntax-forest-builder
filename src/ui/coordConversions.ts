import { PlotCoords } from '../core/types';
import ClientCoords from './ClientCoords';

// Calculate plot coords from client coords & vice versa, when they're different (when zoom and/or pan are implemented)

export const plotCoordsToClientCoords = (plotCoords: PlotCoords): ClientCoords => ({
  clientX: plotCoords.plotX,
  clientY: plotCoords.plotY,
});

export const clientCoordsToPlotCoords = (clientCoords: ClientCoords): PlotCoords => ({
  plotX: clientCoords.clientX,
  plotY: clientCoords.clientY,
});
