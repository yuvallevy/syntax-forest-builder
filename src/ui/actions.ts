import { union } from '../core/objTransforms';
import { Id } from '../core/types';
import { UiState } from './state';

type UiAction =
  | { type: 'selectNodes', plotId: Id, treeIds: Id[], nodeIds: Id[], mode: 'set' | 'add' }
;

export const reducer = (state: UiState, action: UiAction): UiState => {
  switch (action.type) {
    case 'selectNodes':
      return {
        ...state,
        activePlotId: action.plotId,
        selectedTreeIds: action.mode === 'add' ? union(state.selectedTreeIds, action.treeIds) : action.treeIds,
        selectedNodeIds: action.mode === 'add' ? union(state.selectedNodeIds, action.nodeIds) : action.nodeIds,
      };
    default:
      return state;
  }
};
