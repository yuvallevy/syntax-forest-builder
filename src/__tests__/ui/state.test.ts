import { UiAction, UiState, undoableReducer, UndoableUiState } from '../../ui/state';

describe('UI state', () => {
  const initialState: UiState = {
    plots: {},
    activePlotId: 'abc',
    selectedNodes: [{ treeId: 'pqr', nodeId: 'wxy' }],
  };

  const undoableInitialState: UndoableUiState = {
    current: initialState,
    undoStack: [],
    redoStack: [],
  };

  const testCases: [string, UiAction][] = [
    [
      'selects one node replacing current selection',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'xyz' }], mode: 'set' }
    ],
    [
      'selects one node adding to current selection',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'xyz' }], mode: 'add' }
    ],
    [
      'selects two nodes replacing current selection',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'vwx' }, { treeId: 'pqr', nodeId: 'xyz' }], mode: 'set' }
    ],
    [
      'selects two nodes adding to current selection',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'vwx' }, { treeId: 'pqr', nodeId: 'xyz' }], mode: 'add' }
    ],
    [
      'selects two nodes replacing current selection where one node is already selected',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'wxy' }, { treeId: 'pqr', nodeId: 'xyz' }], mode: 'set' }
    ],
    [
      'selects two nodes adding to current selection where one node is already selected',
      { type: 'selectNodes', plotId: 'abc', nodes: [{ treeId: 'pqr', nodeId: 'wxy' }, { treeId: 'pqr', nodeId: 'xyz' }], mode: 'add' }
    ],
  ];

  it.each(testCases)('%s', (name, action) => expect(undoableReducer(undoableInitialState, action).current).toMatchSnapshot());
});