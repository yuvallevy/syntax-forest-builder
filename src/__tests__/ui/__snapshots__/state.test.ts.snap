// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI state selects one node adding to current selection 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "wxy",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;

exports[`UI state selects one node replacing current selection 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;

exports[`UI state selects two nodes adding to current selection 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "wxy",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "vwx",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;

exports[`UI state selects two nodes adding to current selection where one node is already selected 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "wxy",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "wxy",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;

exports[`UI state selects two nodes replacing current selection 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "vwx",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;

exports[`UI state selects two nodes replacing current selection where one node is already selected 1`] = `
Object {
  "activePlotId": "abc",
  "plots": Object {},
  "selectedNodes": Array [
    Object {
      "nodeId": "wxy",
      "treeId": "pqr",
    },
    Object {
      "nodeId": "xyz",
      "treeId": "pqr",
    },
  ],
}
`;
